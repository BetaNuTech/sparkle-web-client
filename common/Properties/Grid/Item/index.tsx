/* eslint-disable no-unused-vars */
import { FunctionComponent, useRef } from 'react';
import styles from './styles.module.scss';
import useVisibility from '../../../hooks/useVisibility';
import propertyModel from '../../../models/property';
import LinkFeature from '../../../LinkFeature';
import TeamValues from '../../../TeamValues';
import features from '../../../../config/features';
import getLastInspectionEntry from '../../../../features/Properties/utils/getLastInspectionEntry';

interface Props {
  property: propertyModel;
  forceVisible?: boolean;
}

const Item: FunctionComponent<Props> = ({ property, forceVisible }) => {
  const lastInspectionEntry = getLastInspectionEntry(property);
  const placeholderRef = useRef(null);
  const { isVisible } = useVisibility(placeholderRef, {}, forceVisible);
  return (
    <li
      ref={placeholderRef}
      className={styles.item}
      data-testid="property-item"
      data-property={property.id}
    >
      {isVisible && (
        <>
          <LinkFeature
            href={`/properties/${property.id}`}
            className={styles.item__wrapper}
            featureEnabled={features.supportBetaPropertyProfile}
          >
            {/* Profile Picture */}
            <aside
              className={styles.item__image}
              style={
                property.photoURL
                  ? {
                      backgroundImage: `url(${property.photoURL})`
                    }
                  : {}
              }
            />

            {/* Profile Description */}
            <div className={styles.item__main}>
              <header>
                <h6
                  className={styles.item__heading}
                  data-testid="property-name"
                  title={property.name}
                >
                  {property.name}
                </h6>
                {property.addr1 && (
                  <p
                    className={styles.item__subHeading}
                    data-testid="property-addr1"
                    title={property.addr1}
                  >
                    {property.addr1}
                  </p>
                )}
              </header>
              {(property.city || property.state) && (
                <p className={styles.item__description}>
                  {/** Do not show city if not present */}
                  {property.city && (
                    <span data-testid="property-city">{`${property.city}, `}</span>
                  )}
                  {/** Do not show state if not present */}
                  {property.state && (
                    <span data-testid="property-state">{`${property.state} `}</span>
                  )}
                  {property.zip && (
                    <span data-testid="property-zip">{property.zip}</span>
                  )}
                </p>
              )}
              {lastInspectionEntry && (
                <p className={styles.item__description}>
                  {lastInspectionEntry}
                </p>
              )}
            </div>
          </LinkFeature>

          {/* Metadata And Edit Button */}
          <footer className={styles.item__footer}>
            <LinkFeature
              href={`/properties/${property.id}/deficient-items`}
              className={styles.item__link}
              featureEnabled={features.supportBetaPropertyDeficient}
            >
              <span className="-mr-sm">Deficient Items</span>
              <TeamValues
                numOfDeficientItems={property.numOfDeficientItems}
                numOfFollowUpActionsForDeficientItems={
                  property.numOfFollowUpActionsForDeficientItems
                }
                numOfRequiredActionsForDeficientItems={
                  property.numOfRequiredActionsForDeficientItems
                }
                isNarrowField={false}
              />
            </LinkFeature>

            <LinkFeature
              href={
                features.supportBetaPropertyUpdate
                  ? `/properties/edit/${property.id}`
                  : `/properties/update/${property.id}`
              }
              className={styles.item__editButton}
              featureEnabled={features.supportBetaPropertyUpdate}
            >
              Edit
            </LinkFeature>
          </footer>
        </>
      )}
    </li>
  );
};

Item.defaultProps = {
  forceVisible: false
};

export default Item;
