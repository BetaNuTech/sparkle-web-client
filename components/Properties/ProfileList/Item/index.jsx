/* eslint-disable no-unused-vars */
import Link from 'next/link';
import PropTypes from 'prop-types';
import styles from './Item.module.scss';
import { TeamValues } from '../../../common/TeamValues';

export const Item = ({ property }) => (
  <li className={styles.item}>
    <Link href="/properties">
      <a className={styles.item__wrapper}>
        {/* Profile Picture */}
        <aside
          className={styles.item__image}
          style={{ backgroundImage: `url(${property.backgroundImage})` }}
        />

        {/* Profile Description */}
        <div className={styles.item__main}>
          <header>
            <h6 className={styles.item__heading}>{property.name}</h6>
            <p className={styles.item__subHeading}>{property.city}</p>
          </header>
          <p className={styles.item__description}>{property.state}</p>
        </div>
      </a>
    </Link>

    {/* Metadata And Edit Button */}
    <footer className={styles.item__footer}>
      <Link href="/edit">
        <a className={styles.item__link}>
          Deficient Items
          <TeamValues
            numOfDeficientItems={property.numOfDeficientItems}
            numOfFollowUpActionsForDeficientItems={
              property.numOfFollowUpActionsForDeficientItems
            }
            numOfRequiredActionsForDeficientItems={
              property.numOfRequiredActionsForDeficientItems
            }
            isNarrowField={false}
          />
        </a>
      </Link>

      <Link href="/edit">
        <a className={styles.item__editButton}>Edit</a>
      </Link>
    </footer>
  </li>
);

Item.propTypes = {
  property: PropTypes.shape({
    backgroundImage: PropTypes.string.isRequired,
    name: PropTypes.string.isRequired,
    addr1: PropTypes.string.isRequired,
    addr2: PropTypes.string.isRequired,
    city: PropTypes.string.isRequired,
    state: PropTypes.string.isRequired,
    lastInspectionDate: PropTypes.number.isRequired,
    lastInspectionScore: PropTypes.number.isRequired
  }).isRequired
};
