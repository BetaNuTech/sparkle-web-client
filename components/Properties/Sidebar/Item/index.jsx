import Link from 'next/link';
import PropTypes from 'prop-types';
import styles from './Item.module.scss';
import { TeamValues } from '../../../common/TeamValues';
import ActionsIcon from '../../../../public/icons/ios/actions.svg';

export const Item = ({ name }) => (
  <li className={styles.item}>
    {/* Team Name */}
    <Link href="/teams">
      <a className={styles.item__teamName}>{name}</a>
    </Link>

    <button aria-label="Open menu" className={styles.item__menuButton}>
      <ActionsIcon />
    </button>

    {/* Metadata */}
    <TeamValues
      numOfDeficientItems={1}
      numOfFollowUpActionsForDeficientItems={22}
      numOfRequiredActionsForDeficientItems={333}
      isNarrowField
    />
  </li>
);

Item.propTypes = {
  name: PropTypes.string.isRequired
};
