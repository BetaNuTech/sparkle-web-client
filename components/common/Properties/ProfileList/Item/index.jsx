import Link from 'next/link';
import PropTypes from 'prop-types';
import styles from './Item.module.scss';
import { TeamValues } from '../../../../shared/TeamValues';

export const Item = ({ backgroundImage, name, address1, address2 }) => (
  <li className={styles.item}>
    <Link href="/properties">
      <a className={styles.item__wrapper}>
        {/* Profile Picture */}
        <aside
          className={styles.item__image}
          style={{ backgroundImage: `url('${backgroundImage}')` }}
        />

        {/* Profile Description */}
        <div className={styles.item__main}>
          <header>
            <h6 className={styles.item__heading}>{name}</h6>
            <p className={styles.item__subHeading}>{address1}</p>
          </header>
          <p className={styles.item__description}>{address2}</p>
        </div>
      </a>
    </Link>

    {/* Metadata And Edit Button */}
    <footer className={styles.item__footer}>
      <Link href="/edit">
        <a className={styles.item__link}>
          Deficient Items
          <TeamValues
            numOfDeficientItems={2}
            numOfFollowUpActionsForDeficientItems={2}
            numOfRequiredActionsForDeficientItems={4}
            isNarrowField={false}
          />
        </a>
      </Link>

      <Link href="/edit">
        <a className={styles.item__editButton}>Edit</a>
      </Link>
    </footer>
  </li>
);

Item.propTypes = {
  backgroundImage: PropTypes.string.isRequired,
  name: PropTypes.string.isRequired,
  address1: PropTypes.string.isRequired,
  address2: PropTypes.string.isRequired
};
