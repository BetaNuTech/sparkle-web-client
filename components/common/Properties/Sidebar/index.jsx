import styles from './Sidebar.module.scss';
import { Item } from './Item';

export const Sidebar = () => (
  <nav className={styles.sidebar}>
    <h4 className={styles.sidebar__heading}>Teams</h4>

    <ul className={styles.sidebar__list}>
      <Item name="Team one" />
      <Item name="Team two" />
      <Item name="Team three" />
    </ul>
  </nav>
);
