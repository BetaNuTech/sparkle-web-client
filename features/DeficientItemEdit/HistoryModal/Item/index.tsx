import { FunctionComponent } from 'react';
import clsx from 'clsx';
import dateUtils from '../../../../common/utils/date';
import HistoryItemDetails from '../ItemDetails';
import UserModel from '../../../../common/models/user';
import { getUserFullname } from '../../../../common/utils/user';

import styles from './styles.module.scss';

interface Props {
  history: any;
  historyType: string;
  isLoading: boolean;
  user: UserModel;
}

const HistoryItem: FunctionComponent<Props> = ({
  history,
  historyType,
  user,
  isLoading
}) => {
  let userName = 'SYSTEM';
  let userEmail = '';

  if (history.user) {
    if (user) {
      userName = getUserFullname(user);
      userEmail = user.email ? `(${user.email})` : '';
    } else {
      userName = 'Unknown';
    }
  }

  return (
    <li className={styles.item}>
      <header>
        <h6
          className={clsx(
            styles.heading,
            isLoading && styles['heading--loading']
          )}
          data-testid="history-user"
        >
          {userName} {userEmail}
        </h6>
        <small className={styles.date} data-testid="history-created-date-time">
          {dateUtils.toUserFullDateDisplay(history.createdAt)} at{' '}
          {dateUtils.toUserTimeDisplay(history.createdAt)}
        </small>
      </header>
      <div className={styles.details}>
        <HistoryItemDetails history={history} historyType={historyType} />
      </div>
    </li>
  );
};

export default HistoryItem;
