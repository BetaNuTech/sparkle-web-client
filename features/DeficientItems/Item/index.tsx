import { FunctionComponent } from 'react';
import LinkFeature from '../../../common/LinkFeature';
import deficientItemModel from '../../../common/models/deficientItem';
import features from '../../../config/features';

interface Props {
  deficientItem: deficientItemModel;
}

const Item: FunctionComponent<Props> = ({ deficientItem }) => (
  <li>
    <LinkFeature
      href={`/properties/${deficientItem.property}/deficient-items/edit/${deficientItem.id}`}
      legacyHref={`/properties/${deficientItem.property}/deficient-items/${deficientItem.id}`}
      featureEnabled={features.supportBetaDeficientItemEdit}
    >
      {deficientItem.id}
    </LinkFeature>
  </li>
);

export default Item;
