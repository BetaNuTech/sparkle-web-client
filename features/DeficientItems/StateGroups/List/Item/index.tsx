import clsx from 'clsx';
import { FunctionComponent, useMemo, useRef } from 'react';
import Router from 'next/router';
import moment from 'moment';
import deficientItemModel from '../../../../../common/models/deficientItem';
import getResponsibilityGroup from '../../../../../common/utils/deficientItem/getResponsibilityGroup';
import features from '../../../../../config/features';
import dateUtils from '../../../../../common/utils/date';
import SelectionIcon from './SelectionIcon';
import useVisibility from '../../../../../common/hooks/useVisibility';

import styles from './styles.module.scss';

interface Props {
  deficientItem: deficientItemModel;
  forceVisible?: boolean;
  isMobile: boolean;
  onSelectDeficiency(state: string, deficiencyId: string): void;
  selectedDeficiencies: string[];
}

const DeficientItemsStateGroupsListItem: FunctionComponent<Props> = ({
  deficientItem,
  forceVisible,
  isMobile,
  onSelectDeficiency,
  selectedDeficiencies
}) => {
  const placeholderRef = useRef(null);

  const { isVisible } = useVisibility(placeholderRef, {}, forceVisible);

  const isClosed = deficientItem.state === 'closed';

  // Determine if DI is past due
  const isPastDue = useMemo(() => {
    const finalDueDate =
      deficientItem.currentDeferredDate || deficientItem.currentDueDate || null;
    const currentDateUnix = moment().unix();

    if (isClosed) {
      return false;
    }

    if (finalDueDate) {
      return finalDueDate < currentDateUnix;
    }
    return false;
  }, [deficientItem, isClosed]);

  const deferredDate = (
    <>
      Deferred:{' '}
      <span className={clsx(isPastDue && !isMobile && '-c-alert')}>
        {dateUtils.toUserDateDisplayWithFullYear(
          deficientItem.currentDeferredDate
        )}
      </span>
    </>
  );

  const dueDate = (
    <>
      Due:{' '}
      <span className={clsx(isPastDue && !isMobile && '-c-alert')}>
        {dateUtils.toUserDateDisplayWithFullYear(deficientItem.currentDueDate)}
      </span>
    </>
  );

  const rightBadgeText =
    (isClosed && deficientItem.isDuplicate && 'Duplicate') ||
    (deficientItem.currentDeferredDate && deferredDate) ||
    (deficientItem.currentDueDate ? dueDate : 'Due: Not Set');

  const onClickItem = () => {
    if (features.supportBetaDeficientItemEdit) {
      Router.push(
        `/properties/${deficientItem.property}/deficient-items/edit/${deficientItem.id}`
      );
    } else {
      window.location.href = `/properties/${deficientItem.property}/deficient-items/${deficientItem.id}`;
    }
  };

  return (
    <li ref={placeholderRef} className={styles.container}>
      {isVisible && (
        <div className={styles.main} onClick={onClickItem}>
          <div
            className={clsx(styles.row, !isMobile && styles['row--reverse'])}
          >
            {!isMobile && (
              <span className={styles.icon}>
                <SelectionIcon
                  itemMainInputType={deficientItem.itemMainInputType}
                  itemMainInputSelection={deficientItem.itemMainInputSelection}
                />
              </span>
            )}
            <div className={styles.row__left}>
              <span
                className={clsx(
                  styles.badge,
                  styles[`badge--${deficientItem.currentResponsibilityGroup}`],
                  isClosed && isMobile && '-o-50'
                )}
                data-testid="current-responsibility-group"
              >
                {getResponsibilityGroup(
                  deficientItem.currentResponsibilityGroup
                ) || 'Not Set'}
              </span>
            </div>
            <div className={styles.row__right}>
              <span
                className={clsx(
                  styles.badge,
                  styles[`badge--${deficientItem.state}`]
                )}
                data-testid="item-right-badge"
              >
                {rightBadgeText}
              </span>
            </div>
          </div>

          <div className={styles.row}>
            {isMobile && (
              <div className={styles.row__left}>
                <span className={clsx(styles.icon, isClosed && '-o-50')}>
                  <SelectionIcon
                    itemMainInputType={deficientItem.itemMainInputType}
                    itemMainInputSelection={
                      deficientItem.itemMainInputSelection
                    }
                  />
                </span>
              </div>
            )}
            <div className={styles.row__right}>
              <span
                className={clsx(
                  styles.title,
                  '-mb-sm',
                  isClosed && isMobile && '-c-gray-light'
                )}
                data-testid="item-title"
              >
                {!isMobile && !isClosed && (
                  <input
                    type="checkbox"
                    checked={
                      selectedDeficiencies.indexOf(deficientItem.id) > -1
                    }
                    data-testid="item-checkbox"
                    className="-mr"
                    onChange={() =>
                      onSelectDeficiency(deficientItem.state, deficientItem.id)
                    }
                    onClick={(e) => e.stopPropagation()}
                  />
                )}
                {deficientItem.itemTitle}
              </span>
              <span
                className={clsx(
                  styles.title,
                  isClosed && isMobile && '-c-info',
                  !isClosed && isMobile && '-c-primary-dark',
                  !isMobile && styles.sectionTitle
                )}
                data-testid="section-title"
              >
                {deficientItem.sectionTitle}
                {deficientItem.sectionSubtitle && !isMobile && (
                  <span data-testid="section-sub-title">
                    {' '}
                    / {deficientItem.sectionSubtitle}
                  </span>
                )}
              </span>
              {deficientItem.sectionSubtitle && isMobile && (
                <span
                  className={clsx(
                    styles.title,
                    isClosed ? '-c-info' : '-c-primary-dark',
                    !isMobile && styles.sectionTitle
                  )}
                  data-testid="section-sub-title"
                >
                  {deficientItem.sectionSubtitle}
                </span>
              )}
            </div>
          </div>
        </div>
      )}
    </li>
  );
};

export default DeficientItemsStateGroupsListItem;
