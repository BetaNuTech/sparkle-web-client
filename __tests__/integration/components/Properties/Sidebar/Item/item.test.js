import React from 'react';
import { shallow } from 'enzyme';
import { Item } from '../../../../../../components/Properties/Sidebar/Item';

describe('Item component', () => {
  describe('rendering', () => {
    let wrapper;
    beforeEach(() => {
      const props = {
        name: 'teamOne'
      };
      wrapper = shallow(<Item {...props} />);
    });

    it('renders correctly', () => {
      expect(wrapper).toMatchSnapshot();
    });

    it('should render name', () => {
      const name = wrapper.find('a').text();
      expect(name).toBe('teamOne');
    });
  });
});
