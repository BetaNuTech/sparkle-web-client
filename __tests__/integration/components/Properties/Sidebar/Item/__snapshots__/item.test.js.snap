// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Item component rendering renders correctly 1`] = `ShallowWrapper {}`;
